futures-util/src/stream/try_stream/try_unfold.rs

```diff
@@ -0,0 +1,144 @@
+/// `TryStream::try_fold()` reduces a `TryStream` to one single value,
+/// `try_unfold()` creates a `TryStream` from a seed value.
+///
+/// `try_unfold()` will call the provided closure with the provided seed, then
+/// wait for the returned `TryFuture` to complete with `(a, b)`. It will then
+/// yield the value `a`, and use `b` as the next internal state.
+///
+/// If the closure returns `None` instead of `Some(TryFuture)`, then the
+/// `try_unfold()` will stop producing items and return `Poll::Ready(None)` in
+/// future calls to `poll()`.
+///
+/// In case of error generated by the returned `TryFuture`, the error will be
+/// returned by the `TryStream`. The `TryStream` will then yield
+/// `Poll::Ready(None)` in future calls to `poll()`.
+///
+/// This function can typically be used when wanting to go from the "world of
+/// futures" to the "world of streams": the provided closure can build a
+/// `TryFuture` using other library functions working on futures, and
+/// `try_unfold()` will turn it into a `TryStream` by repeating the operation.
+///
+/// # Example
+///
+/// ```
+/// # #[derive(Debug, PartialEq)]
+/// # struct SomeError;
+/// # futures::executor::block_on(async {
+/// use futures::stream::{self, TryStreamExt};
+///
+/// let stream = stream::try_unfold(0, |state| async move {
+///     if state < 0 {
+///         return Err(SomeError);
+///     }
+///
+///     if state <= 2 {
+///         let next_state = state + 1;
+///         let yielded = state * 2;
+///         Ok(Some((yielded, next_state)))
+///     } else {
+///         Ok(None)
+///     }
+/// });
+///
+/// let result: Result<Vec<i32>, _> = stream.try_collect().await;
+/// assert_eq!(result, Ok(vec![0, 2, 4]));
+/// # });
+/// ```
+pub fn try_unfold<T, F, Fut, Item>(init: T, f: F) -> TryUnfold<T, F, Fut>
+where
+    F: FnMut(T) -> Fut,
+    Fut: TryFuture<Ok = Option<(Item, T)>>,
+{
+    TryUnfold {
+        f,
+        state: Some(init),
+        fut: None,
+    }
+}
+
+/// Stream for the [`try_unfold`] function.
+#[must_use = "streams do nothing unless polled"]
+pub struct TryUnfold<T, F, Fut> {
+    f: F,
+    state: Option<T>,
+    fut: Option<Fut>,
+}
+
+impl<T, F, Fut: Unpin> Unpin for TryUnfold<T, F, Fut> {}
+
+impl<T, F, Fut> fmt::Debug for TryUnfold<T, F, Fut>
+where
+    T: fmt::Debug,
+    Fut: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TryUnfold")
+            .field("state", &self.state)
+            .field("fut", &self.fut)
+            .finish()
+    }
+}
+
+impl<T, F, Fut> TryUnfold<T, F, Fut> {
+    unsafe_unpinned!(f: F);
+    unsafe_unpinned!(state: Option<T>);
+    unsafe_pinned!(fut: Option<Fut>);
+}
+
+impl<T, F, Fut, Item> FusedStream for TryUnfold<T, F, Fut>
+where
+    F: FnMut(T) -> Fut,
+    Fut: TryFuture<Ok = Option<(Item, T)>>,
+{
+    fn is_terminated(&self) -> bool {
+        self.state.is_none() && self.fut.is_none()
```

Comment:
This will yield `true` even before `None` has been returned in the case where the future errors, which is incorrect.

Reply:
Why is that incorrect? The docs for `FusedStream` say that `is_terminated` returns true if the stream should no longer be polled, which is usually the case after `None` is returned but may also be the case without that. When we know that all future calls to `poll_next` will return `None`, is it not reasonable to return `true` from `.is_terminated()`?

Reply:
No, `is_terminated()` should only return `true` once `None` has been yielded.

Reply:
Okay :man_shrugging: